packages/base-controller/src/ControllerMessenger.ts

```diff
@@ -41,13 +61,19 @@ export type SelectorEventHandler<SelectorReturnValue> = (
 
 export type ActionConstraint = {
   type: string;
-  handler: (...args: any) => unknown;
+  handler: ((...args: never) => unknown) | ((...args: never[]) => unknown);
+};
```

Comment:
So it turns out that `(...args: never[]) => unknown` probably isn't the universal supertype for functions.

The error messages @Gudahtt linked here https://github.com/MetaMask/core/pull/1890#discussion_r1391617360 reflect the fact that in general, array types of arbitrary length are not assignable to tuple types that have fixed types designated to each index position (since the array is wider than the tuple).

This means that there's a whole subcategory of functions that `(...args: never[]) => unknown` doesn't cover: functions that have a finite number of parameters that are strictly ordered and typed, making their argument list evaluate as a fixed-length or variadic **tuple** type, instead of an expandable **array** type.

That is, `(...args: never[]) => unknown` covers the cases where all of the params can be typed together as a group e.g. `(...args: (string | number)[]) => T`. But to cover cases like `(a: string, b: number, ...rest: unknown[]) => T`, we need another top type with an even narrower type for the params: `(...args: never) => unknown`. 

In short, I think the top type that includes both of these function categories is: `((...args: never) => unknown) | ((...args: never[]) => unknown)`.

(`| never` in general is analogous to `+ 0`, so `(...args: (never[] | never)) => unknown` doesn't work.)

Reply:
Great sleuthing! 👏🏻